File: app/src/main/java/com/mgaetan89/showsrage/adapter/LogsAdapter.java
package com.mgaetan89.showsrage.adapter;

import android.databinding.DataBindingUtil;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.mgaetan89.showsrage.R;
import com.mgaetan89.showsrage.databinding.AdapterLogsListBinding;
import com.mgaetan89.showsrage.model.LogEntry;
import com.mgaetan89.showsrage.presenter.LogPresenter;

import java.util.Collections;
import java.util.List;

public class LogsAdapter extends RecyclerView.Adapter<LogsAdapter.ViewHolder> {
	@NonNull
	private List<String> logs = Collections.emptyList();

	public LogsAdapter(@Nullable List<String> logs) {
		if (logs == null) {
			this.logs = Collections.emptyList();
		} else {
			this.logs = logs;
		}
	}

	@Override
	public int getItemCount() {
		return this.logs.size();
	}

	@Override
	public void onBindViewHolder(ViewHolder holder, int position) {
		LogEntry logEntry = new LogEntry(this.logs.get(position));

		holder.bind(new LogPresenter(logEntry));
	}

	@Override
	public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
		View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.adapter_logs_list, parent, false);

		return new ViewHolder(view);
	}

	public static class ViewHolder extends RecyclerView.ViewHolder {
		private final AdapterLogsListBinding binding;

		public ViewHolder(View view) {
			super(view);

			this.binding = DataBindingUtil.bind(view);
		}

		public void bind(LogPresenter logEntry) {
			this.binding.setLog(logEntry);
		}
	}
}

File: app/src/main/kotlin/com/mgaetan89/showsrage/adapter/LogsAdapter.kt
package com.mgaetan89.showsrage.adapter

import android.support.v4.content.ContextCompat
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.mgaetan89.showsrage.R
import com.mgaetan89.showsrage.model.LogEntry
import com.mgaetan89.showsrage.presenter.LogPresenter
import io.realm.RealmRecyclerViewAdapter
import io.realm.RealmResults
import kotlinx.android.synthetic.main.adapter_logs_list.view.log_date_time
import kotlinx.android.synthetic.main.adapter_logs_list.view.log_error_type
import kotlinx.android.synthetic.main.adapter_logs_list.view.log_group
import kotlinx.android.synthetic.main.adapter_logs_list.view.log_message

class LogsAdapter(logs: RealmResults<LogEntry>) : RealmRecyclerViewAdapter<LogEntry, LogsAdapter.ViewHolder>(logs, true) {
	override fun onBindViewHolder(holder: ViewHolder, position: Int) {
		val logEntry = this.getItem(position)?.takeIf { it.isValid } ?: return

		holder.bind(logEntry)
	}

	override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
		val view = LayoutInflater.from(parent.context).inflate(R.layout.adapter_logs_list, parent, false)

		return ViewHolder(view)
	}

	class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
		private val dateTime = view.log_date_time
		private val errorType = view.log_error_type
		private val group = view.log_group
		private val message = view.log_message

		init {
			this.group.isSelected = true
		}

		fun bind(logEntry: LogEntry) {
			val context = this.itemView.context
			val presenter = LogPresenter(logEntry)

			this.dateTime.text = presenter.getDateTime()
			this.errorType.text = presenter.getErrorType()
			this.errorType.setTextColor(ContextCompat.getColor(context, presenter.getErrorColor()))
			this.group.text = presenter.getGroup()
			this.message.text = presenter.getMessage()
		}
	}
}
